### PR TITLE
JetStream scaler should query the stream consumer leader when clustered

### PR DESCRIPTION
**Pull request by rayjanoka**
I went to test drive the new JetStream scaler on my project and it started to have an issue. Eventually I figured out that the NATS monitoring endpoint isn't reporting accurate metrics on stream consumers from all pods in a cluster. 

I found that when running a cluster of nats pods, only the stream consumer's leader pod reports the accurate number of messages in the queue. The jetstream scaler would work fine for me at first as long as keda was connected to the nats consumer leader pod, but as soon as I bounced keda's connection to a nats consumer replica pod the `num_pending` value there counts up but never down, so my deployment just scales up and up and up.

```shell
➜ nats consumer info test-stream durable

Cluster Information:
                Name: nats
              Leader: nats-1   <---- Leader Pod
             Replica: nats-0, current, seen 0.33s ago
             Replica: nats-2, current, seen 0.33s ago
```

```shell
# Consumer Leader (accurate)
➜ curl -s "http://nats-1.nats.nats.svc.cluster.local:8222/jsz?consumers=true" | grep num_pending
              "num_pending": 0,

# Replicas (counting up only)
➜ curl -s "http://nats-0.nats.nats.svc.cluster.local:8222/jsz?consumers=true" | grep num_pending
              "num_pending": 6,
➜ curl -s "http://nats-2.nats.nats.svc.cluster.local:8222/jsz?consumers=true" | grep num_pending
              "num_pending": 6,
```

We are able to discover the jetstream consumer's leader via the existing monitoring endpoint call.

```shell
$ curl -s "http://nats.nats.svc.cluster.local:8222/jsz?consumers=true" | jq '.account_details[0].stream_detail[0].consumer_detail[0].cluster.leader'
"nats-0"
```

To get the accurate count to the scaler I wrote a change to make a 2nd request directly to that consumer leader pod via the headless svc, ex `consumer-leader-pod.nats.nats.svc.cluster.local`.

NOTE: This fix will only work for clusters who have the same number of pods as stream replicas, ex. 3 pods and 3 stream replicas or 5 pods and 5 stream replicas.  

```shell
➜ nats stream info test-stream
Information for Stream test-stream

Configuration:

             Subjects: sub
             Replicas: 3   <---- stream replica setting here
```

this is resolved below https://github.com/kedacore/keda/pull/3564#issuecomment-1320835018
~If there are less stream replicas than pods, ex. 3 pods and only 1 stream replica, NATS will only display metrics for that consumer on the single pod that the stream is assigned to, the other pods have no record of the consumer at all so it will only find the metric if the k8s svc happens to terminate you to that particular pod. Without any other clues to figure out which pod has the metric I think we'd have to rotate through each NATS pod blindly until we found the pod, not great. (I can document this limitation in the keda-docs for now)~

@goku321 for visibility - thanks for contributing this!

I'll send a ticket over to the NATS side as well and see if they can work to provide accurate metrics across all NATS servers in a cluster so no matter what server we connect to we see everything.

### Checklist

- [X] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))
- [X] Tests have been added
- [NA] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [NA] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [X] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)

Fixes https://github.com/kedacore/keda/issues/3860

Relates to #

### CODE REVIEW

**Comment by goku321** [comment]
@rayjanoka Thank you so much for giving it a test drive! Really appreciate all your efforts :)

Now, I'm thinking if we really need the single server JetStream [tests](https://github.com/kedacore/keda/tree/main/tests/scalers/nats_jetstream) and instead add tests for clustered JetStream. What do you think?

**Comment by rayjanoka** [comment]
> @rayjanoka Thank you so much for giving it a test drive! Really appreciate all your efforts :)
> 
> Now, I'm thinking if we really need the single server JetStream [tests](https://github.com/kedacore/keda/tree/main/tests/scalers/nats_jetstream) and instead add tests for clustered JetStream. What do you think?

@goku321 nice! I did notice we had separate tests like that for redis and I think it is probably a good idea here as well. I don't have much experience writing tests, and it might take me a while to circle back to this to give those a shot. If you or anyone else would like to take a crack at it that would be fine.

**Comment by goku321** [comment]
@rayjanoka Sure. I can start with the tests in a week or so. I hope that's okay :)

File: pkg/scalers/nats_jetstream_scaler.go [hunk 1]
```diff
@@ -241,6 +302,45 @@ func (s *natsJetStreamScaler) GetMetrics(ctx context.Context, metricName string,
...
+		}
+
+		// Find and assign the leader stream that we are looking for.
+		for _, account := range jsAccountLeaderResp.Accounts {
+			if account.Name == s.metadata.account {
+				for _, stream := range account.Streams {
+					if stream.Name == s.metadata.stream {
+						s.leaderStream = stream
+					}
+				}
```
Comment (JorTurFer) [inline pkg/scalers/nats_jetstream_scaler.go:343]: maybe we should extract this code into a function because it's duplicated

Comment (rayjanoka) [reply pkg/scalers/nats_jetstream_scaler.go:343]: ya...I didn't want to upend the structure, but the only difference between the duplicate sections are the return values.

Comment (rayjanoka) [reply pkg/scalers/nats_jetstream_scaler.go:343]: yeah, probably a good idea, the only difference between the duplicate sections are the return values. maybe @goku321 will have some input.

Comment (goku321) [reply pkg/scalers/nats_jetstream_scaler.go:343]: Hi @rayjanoka I know I've already delayed this so much but I've been dealing with some health issues recently and couldn't spend any time on the tests. Allow me some time and I'll try to speed things up. Really appreciate your patience.

Comment (rayjanoka) [reply pkg/scalers/nats_jetstream_scaler.go:343]: take care of yourself @goku321, no rush here.

Comment (zroubalik) [reply pkg/scalers/nats_jetstream_scaler.go:343]: @goku321 absolutely, take care of yourself first

**Comment by rayjanoka** [comment]
I'm back on this, I'm going to work on putting this in a function and writing some tests.

**Comment by rayjanoka** [comment]
I took a look at the tests, and I think we're already covering everything that something similar, like redis, is testing for. I think I'd need to run a `nats-server` during the test to test the other functions. I'm open to giving that a shot, but if not this is good to review and ship.

**Comment by rayjanoka** [comment]
ok, this is looking good! I did some serious surgery and we're at 96% test coverage! 😆 

I'm going to give this a run in my lab and I'll let ya know how it goes.

File: pkg/scalers/nats_jetstream_scaler.go (hunk 1 above)
Comment (rayjanoka) [reply pkg/scalers/nats_jetstream_scaler.go]: this is resolved @JorTurFer

**Comment by rayjanoka** [comment]
I think this is ready for review, it tested ok in my environment.
* De-duplicated code
* Lots of new tests
* All tests passed! coverage: image

File: CHANGELOG.md [hunk 2]
```diff
@@ -36,6 +36,10 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - [v1.1.0](#v110)
 - [v1.0.0](#v100)
 
+## Unreleased
+
+- **JetStream:** JetStream scaler use stream consumer leader metric when clustered ([#2391](https://github.com/kedacore/keda/issues/2391))
```
Comment (zroubalik) [inline CHANGELOG.md:41]: Please move this to `### Improvements` section (alphabetically sorted) and please name the scaler `**Nats Jetstream**`.

**Review by zroubalik — COMMENTED** [review]
This change is looking good, there are a few leftovers:

- could you please create a new issue describing the problem and reference it in the changelog?
- the test coverage is great! I think that @goku321 was talking about e2e test that we have, to support clustered setup: https://github.com/kedacore/keda/issues/2391

File: CHANGELOG.md (hunk 2 above)
Comment (rayjanoka) [reply CHANGELOG.md]: all set on this, thanks!

Comment (rayjanoka) [reply CHANGELOG.md]: I see where the e2e test is now...I can probably add a clustered test...should I do that here or in another PR?

Comment (zroubalik) [reply CHANGELOG.md]: Please do it here, thanks!

File: CHANGELOG.md [hunk 3]
```diff
@@ -64,6 +64,7 @@ To learn more about active deprecations, we recommend checking [GitHub Discussio
 - **Prometheus Scaler:** Introduce skipping of certificate check for unsigned certs ([#2310](https://github.com/kedacore/keda/issues/2310))
 - **Pulsar Scaler:** Add support for bearer token and basic auth ([#3844](https://github.com/kedacore/keda/issues/3844))
 - **Pulsar Scaler:** Add support for partitioned topics ([#3833](https://github.com/kedacore/keda/issues/3833))
+- **NATS Jetstream Scaler:** Query the stream consumer leader when clustered ([#3860](https://github.com/kedacore/keda/issues/3860))
```
Comment (zroubalik) [inline CHANGELOG.md:67]: Please sort alphabetically.

Comment (rayjanoka) [reply CHANGELOG.md]: got it

**Comment by rayjanoka** [comment]
Updates
* clustered e2e test with 3 nats nodes - takes about 6.5 mins to run
  * test all combinations of stream replicas (1, 2, & 3) in the cluster
* supporting all the stream replicas required a bit of logic to discover the node. when the stream is only on 1 node we need to just keep trying nodes until we find the stream.

**Comment by JorTurFer** [comment]
> Am I able to trigger the e2e test here? or if one of the crew could start it we can see if it works here.

No you can't. To trigger them, it's required to be in a [specific team in the org](https://github.com/kedacore/keda/blob/main/.github/workflows/pr-e2e.yml#L25). I have already triggered

**Comment by rayjanoka** [comment]
ok, it looks like it worked...I can't see the log, but I see a green check.

I'll take one more look over this on Monday and then we can complete the review. thanks!

**Comment by JorTurFer** [comment]
You can find the specific workflow logs following the link in the comment
image

There, you have to go to the `Execute e2e tests` job:
image

And there you only need to check the step:
image

**Comment by JorTurFer** [comment]
> I'll take one more look over everything on Monday, then we can do the review and get this finished. thanks!

Perfect! Thanks for your contribution!

**Comment by rayjanoka** [comment]
> You can find the specific workflow logs following the link in the comment 
ah hah! of course, the log looks good good!

**Comment by rayjanoka** [comment]
ok, I just rebased, should be good for review and delivery! @goku321 @JorTurFer @zroubalik

**Comment by rayjanoka** [comment]
> /run-e2e nats* **Update:** You can check the progress [here](https://github.com/kedacore/keda/actions/runs/3572087858)

LGTM, the clustered test is long, but I helped a bit overall by cutting 3.5 mins off the standalone test by pushing less messages through in the test, 1000 -> 300.

standalone e2e test: 59 secs
clustered e2e test: 7.3 mins

**Comment by tomkerkhove** [comment]
@JorTurFer @zroubalik Can you re-review this PR please to see if we need to change things before our release on Thursday?
@rayjanoka can you fix the merge conflict please?

**Comment by rayjanoka** [comment]
> @JorTurFer @zroubalik Can you re-review this PR please to see if we need to change things before our release on Thursday? @rayjanoka can you fix the merge conflict please?

fixed! thanks @tomkerkhove!

**Review by zroubalik — APPROVED** [review]
LGTM

thanks a lot for the contribution!

**Pull request merged**